README.md
# chef-docker [![Build Status](https://secure.travis-ci.org/bflad/chef-docker.png?branch=master)](http://travis-ci.org/bflad/chef-docker)
[![Gitter](https://badges.gitter.im/Join Chat.svg)](https://gitter.im/bflad/chef-docker?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

## Description

Installs/Configures [Docker](http://docker.io). Please see [COMPATIBILITY.md](COMPATIBILITY.md) for more information about Docker versions that are tested and supported by cookbook versions along with LWRP features.

This cookbook was inspired by @thoward's docker-cookbook: https://github.com/thoward/docker-cookbook

## Breaking Change Alert

In version 1.0 of this cookbook, we will be making a significant breaking changes including the way that we handle the custom resources (`docker_image`, `docker_container` and `docker_registry`). It is highly recommended that you constrain the version of the cookbook you are using in the appropriate places.
  - metadata.rb
  - Chef Environments
  - Berksfile
  - Chef Policyfile

More details about specific changes will be documented in the [1.0_CHANGES.md](1.0_CHANGES.md) file.

## Requirements

### Chef

* Chef 11+

### Platforms

* Amazon 2014.03.1 (experimental)
* CentOS 6
* Debian 7
* Fedora 19, 20
* Mac OS X (only docker installation currently)
* Oracle 6
* RHEL 6
* Ubuntu 12.04, 12.10, 13.04, 13.10, 14.04 (experimental)

### Cookbooks

[Opscode Cookbooks](https://github.com/opscode-cookbooks/)

* [apt](https://github.com/opscode-cookbooks/apt)
* [git](https://github.com/opscode-cookbooks/git)
* [homebrew](https://github.com/opscode-cookbooks/homebrew)
* [yum-epel](https://github.com/opscode-cookbooks/yum-epel)

Third-Party Cookbooks

* [aufs](https://github.com/bflad/chef-aufs)
* [device-mapper](https://github.com/bflad/chef-device-mapper)
* [golang](https://github.com/NOX73/chef-golang)
* [lxc](https://github.com/hw-cookbooks/lxc)
* [modules](https://github.com/Youscribe/modules-cookbook)
* [sysctl](https://github.com/onehealth-cookbooks/sysctl)

## Usage

### Default Installation

* Add `recipe[docker]` to your node's run list

### Execution Drivers

If your system is running a Docker version before 0.9, you'll need to explicitly set up LXC outside of this cookbook. This will likely be true for most distros after Docker 1.0 and chef-docker 1.0 is released.
* [lxc on community site](http://community.opscode.com/cookbooks/lxc)
* [lxc on Github](https://github.com/hw-cookbooks/lxc/)

### Storage Drivers

Beginning in chef-docker 1.0, storage driver installation and configuration is expected to be handled before this cookbook's execution, except where required by Docker.

#### AUFS

If you need AUFS support, consider adding the aufs cookbook to your node/recipe before docker.
* [aufs on community site](http://community.opscode.com/cookbooks/aufs)
* [chef-aufs on Github](https://github.com/bflad/chef-aufs)

Then, set the `storage_driver` attribute of this cookbook to `aufs`.

#### device-mapper

If you need device-mapper support, consider adding the device-mapper cookbook to your node/recipe before docker.
* [device-mapper on community site](http://community.opscode.com/cookbooks/device-mapper)
* [chef-device-mapper on Github](https://github.com/bflad/chef-device-mapper)

Then, set the `storage_driver` attribute of this cookbook to `devicemapper` (please note lack of dash).

### Ubuntu 14.04 Package Installation via Docker PPA

By default, this cookbook will use the docker.io package from Ubuntu 14.04's repository. To use the Docker PPA package, just set the repo_url attribute to the Docker PPA URL. e.g. `node.set['docker']['package']['repo_url'] = 'https://get.docker.io/ubuntu'`

## Attributes

### Installation/System Attributes

These attributes are under the `node['docker']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
arch | Architecture for docker binary (note: Docker only currently supports x86_64) | String | auto-detected (see attributes/default.rb)
group_members | Users to manage in `node['docker']['group']` | Array of Strings | []
init_type | Init type for docker ("runit", "systemd", "sysv", or "upstart") | String | auto-detected (see attributes/default.rb)
install_dir | Installation directory for docker binary (custom setting only valid for non-package installations) | String | auto-detected (see attributes/default.rb)
install_type | Installation type for docker ("binary", "package" or "source") | String | package
ipv4_forward | Sysctl set net.ipv4.ip_forward to 1 | TrueClass, FalseClass | true
ipv6_forward | Sysctl set net.ipv6.conf.all.forwarding to 1 | TrueClass, FalseClass | true
version | Version of docker | String | nil

#### Binary Installation Attributes

These attributes are under the `node['docker']['binary']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
checksum | Optional SHA256 checksum for docker binary | String | auto-detected (see attributes/default.rb)
version | Version of docker binary | String | `node['docker']['version']` (if set) or `latest`
url | URL for downloading docker binary | String | `http://get.docker.io/builds/#{node['kernel']['name']}/#{node['docker']['arch']}/docker-#{node['docker']['binary']['version']}`

#### Package Installation Attributes

These attributes are under the `node['docker']['package']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
action | Action for docker packages ("install", "update", etc.) | String | install
distribution | Distribution for docker packages | String | auto-detected (see attributes/default.rb)
name | Override Docker package name | String | auto-detected (see attributes/default.rb)
repo_url | Repository URL for docker packages | String | auto-detected (see attributes/default.rb)
repo_key | Repository GPG key URL for docker packages | String | https://get.docker.io/gpg

#### Source Installation Attributes

These attributes are under the `node['docker']['source']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
ref | Repository reference for docker source | String | master
url | Repository URL for docker source | String | https://github.com/dotcloud/docker.git

### Docker Daemon Attributes

For more information: http://docs.docker.io/en/latest/reference/commandline/cli/#daemon

These attributes are under the `node['docker']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
api_enable_cors | Enable CORS headers in API | TrueClass, FalseClass | nil
bind_socket (*DEPRECATED*) | Socket path that docker should bind | String | unix:///var/run/docker.sock
bind_uri (*DEPRECATED*) | TCP URI docker should bind | String | nil
bip | Use this CIDR notation address for the network bridge's IP, not compatible with `bridge` | String | nil
bridge | Attach containers to a pre-existing network bridge; use 'none' to disable container networking | String | nil
debug | Enable debug mode | TrueClass, FalseClass | nil (implicitly false)
dns | DNS server(s) for containers | String, Array | nil
dns_search | DNS search domain(s) for containers | String, Array | nil
exec_driver | Execution driver for docker | String | nil (implicitly native as of 0.9.0)
graph | Path to use as the root of the docker runtime | String | nil (implicitly /var/lib/docker)
group | Group for docker socket and group_members | String | nil (implicitly docker)
host | Socket(s) that docker should bind | String, Array | unix:///var/run/docker.sock
http_proxy | HTTP_PROXY environment variable | String | nil
icc | Enable inter-container communication | TrueClass, FalseClass | nil (implicitly true)
insecure-registry | List of well-known insecure registries | String, Array | nil
ip | Default IP address to use when binding container ports | String | nil (implicitly 0.0.0.0)
iptables | Enable Docker's addition of iptables rules | TrueClass, FalseClass | nil (implicitly true)
logfile | Set custom DOCKER_LOGFILE | String | nil
mtu | Set the containers network MTU | Fixnum | nil (implicitly default route MTU or 1500 if no default route is available)
no_proxy | NO_PROXY environment variable | String | nil
options | Additional options to pass to docker. These could be flags like "-api-enable-cors". | String | nil
pidfile | Path to use for daemon PID file | String | nil (implicitly /var/run/docker.pid)
ramdisk | Set DOCKER_RAMDISK when using RAM disk | TrueClass or FalseClass | false
registry-mirror | List of docker registry mirrors | String, Array | nil
restart (*DEPRECATED*) | Restart containers on boot | TrueClass or FalseClass | nil
selinux_enabled | Enable SELinux | TrueClass or FalseClass | nil
storage_driver | Storage driver for docker | String | nil
storage_opt | Storage driver options | String, Array | nil
tls | Use TLS | TrueClass, FalseClass | nil (implicitly false)
tlscacert | Trust only remotes providing a certificate signed by the CA given here | String | nil (implicitly ~/.docker/ca.pem)
tlscert | Path to TLS certificate file | String | nil (implicitly ~/.docker/cert.pem)
tlskey | Path to TLS key file | String | nil (implicitly ~/.docker/key.pem)
tlsverify | Use TLS and verify the remote (daemon: verify client, client: verify daemon) | TrueClass, FalseClass | nil (implicitly false)
tmpdir | TMPDIR environment variable | String | nil

### LWRP Attributes

These attributes are under the `node['docker']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
docker_daemon_timeout | Timeout to wait for the docker daemon to start in seconds for LWRP commands | Fixnum | 20

#### docker_container Attributes

These attributes are under the `node['docker']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
container_cmd_timeout | container LWRP default cmd_timeout seconds | Fixnum | 60
container_init_type | Init type for docker containers (nil, "runit", "systemd", "sysv", or "upstart") | String | `node['docker']['init_type']`

#### docker_image Attributes

These attributes are under the `node['docker']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
image_cmd_timeout | image LWRP default cmd_timeout seconds | Fixnum | 300

#### docker_registry Attributes

These attributes are under the `node['docker']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
registry_cmd_timeout | registry LWRP default cmd_timeout seconds | Fixnum | 60

## Recipes

* `recipe[docker]` Installs/Configures Docker
* `recipe[docker::aufs]` Installs/Loads AUFS Linux module
* `recipe[docker::binary]` Installs Docker binary
* `recipe[docker::cgroups]` Installs/configures default platform Control Groups support
* `recipe[docker::devicemapper]` Installs/Configures Device Mapper
* `recipe[docker::group]` Installs/Configures docker group
* `recipe[docker::lxc]` Installs/configures default platform LXC support
* `recipe[docker::package]` Installs Docker via package
* `recipe[docker::runit]` Installs/Starts Docker via runit
* `recipe[docker::source]` Installs Docker via source
* `recipe[docker::systemd]` Installs/Starts Docker via systemd
* `recipe[docker::sysv]` Installs/Starts Docker via SysV
* `recipe[docker::upstart]` Installs/Starts Docker via Upstart

## LWRPs

* docker_container: container operations
* docker_image: image/repository operations
* docker_registry: registry operations

### Getting Started

Here's a quick example of pulling the latest image and running a container with exposed ports (creates service automatically):

```ruby
# Pull latest image
docker_image 'samalba/docker-registry'

# Run container exposing ports
docker_container 'samalba/docker-registry' do
  detach true
  port '5000:5000'
  env 'SETTINGS_FLAVOR=local'
  volume '/mnt/docker:/docker-storage'
end
```

Maybe you want to automatically update your private registry with changes from your container?

```ruby
# Login to private registry
docker_registry 'https://docker-registry.example.com/' do
  username 'shipper'
  password 'iloveshipping'
end

# Pull tagged image
docker_image 'apps/crowsnest' do
  tag 'not-latest'
end

# Run container
docker_container 'crowsnest'

# Save current timestamp
timestamp = Time.new.strftime('%Y%m%d%H%M')

# Commit container changes
docker_container 'crowsnest' do
  repository 'apps'
  tag timestamp
  action :commit
end

# Push image
docker_image 'crowsnest' do
  repository 'apps'
  tag timestamp
  action :push
end
```

See full documentation for each LWRP and action below for more information.

### docker_container

Below are the available actions for the LWRP, default being `run`.

These attributes are associated with all LWRP actions.

Attribute | Description | Type | Default
----------|-------------|------|--------
cmd_timeout | Timeout for docker commands (catchable exception: `Chef::Provider::Docker::Container::CommandTimeout`)| Integer | `node['docker']['container_cmd_timeout']`
command | Command to run in or identify container | String | nil
container_name | Name for container/service | String | nil

#### docker_container action :commit

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
author | Author for commit | String | nil
message | Message for commit | String | nil
repository | Remote repository | String | nil
run | Configuration to be applied when the image is launched with `docker run` | String | nil
tag | Specific tag for image | String | nil

Commit a container with optional repository, run specification, and tag:

```ruby
docker_container 'myApp' do
  repository 'myRepo'
  tag Time.new.strftime("%Y%m%d%H%M")
  run '{"Cmd": ["cat", "/world"], "PortSpecs": ["22"]}'
  action :commit
end
```

#### docker_container action :cp

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
destination | Host path to copy file | String | nil
source | Container path to get file | String | nil

Copying a file from container to host:

```ruby
docker_container 'myApp' do
  source '/path/to/container/file'
  destination '/path/to/save/on/host'
  action :cp
end
```

#### docker_container action :create

By default, this will handle creating a service for the container when action is create, run or start. `set['docker']['container_init_type'] = false` or add `init_type false` for LWRP to disable this behavior.

Attributes for this action can be found in the `run` action (except for the `detach` attribute).

#### docker_container action :export

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
destination | Host path to save tarball | String | nil

Exporting container to host:

```ruby
docker_container 'myApp' do
  destination '/path/to/save/on/host.tgz'
  action :export
end
```

#### docker_container action :kill

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
cookbook | Cookbook to grab any templates | String | docker
init_type | Init type for container service handling | FalseClass, String | `node['docker']['container_init_type']`
init_template | Template to use for init configuration | String | nil
signal | Signal to send to the container | String | nil (implicitly KILL)
socket_template | Template to use for configuring socket (relevent for init_type systemd only) | String | nil

Kill a running container:

```ruby
docker_container 'shipyard' do
  action :kill
end
```

Send SIGQUIT to a running container:

```ruby
docker_container 'shipyard' do
  signal 'QUIT'
  action :kill
end
```

#### docker_container action :redeploy

Stops, removes, and runs a container. Useful for notifications from image build/pull.

Attributes for this action can be found in the `stop`, `remove`, and `run` actions.

Redeploy container when new image is pulled:

```ruby
docker_image 'shipyard/shipyard' do
  action :pull
  notifies :redeploy, 'docker_container[shipyard]', :immediately
end

docker_container 'shipyard' do
  # Other attributes
  action :run
end
```

#### docker_container action :remove

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
cookbook | Cookbook to grab any templates | String | docker
force | Force removal | TrueClass, FalseClass | nil
init_type | Init type for container service handling | FalseClass, String | `node['docker']['container_init_type']`
init_template | Template to use for init configuration | String | nil
socket_template | Template to use for configuring socket (relevent for init_type systemd only) | String | nil

Remove a container:

```ruby
docker_container 'shipyard' do
  action :remove
end
```

#### docker_container action :remove_link

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
link | Link to remove from container | String | nil

Remove a container:

```ruby
docker_container 'shipyard' do
  link 'foo'
  action :remove_link
end
```

#### docker_container action :remove_volume

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
volume | Volume(s) to remove from container | String, Array | nil

Remove a container:

```ruby
docker_container 'shipyard' do
  volume %w(/extravol1 /extravol2)
  action :remove_volume
end
```

#### docker_container action :restart

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
cookbook | Cookbook to grab any templates | String | docker
init_type | Init type for container service handling | FalseClass, String | `node['docker']['container_init_type']`
init_template | Template to use for init configuration | String | nil
socket_template | Template to use for configuring socket (relevent for init_type systemd only) | String | nil

Restart a container:

```ruby
docker_container 'shipyard' do
  action :restart
end
```

#### docker_container action :run

By default, this will handle creating a service for the container when action is create, run or start. `set['docker']['container_init_type'] = false` or add `init_type false` for LWRP to disable this behavior.

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
additional_host | Add a custom host-to-IP mapping (host:ip) | String, Array | nil
attach | Attach container's stdout/stderr and forward all signals to the process | TrueClass, FalseClass | nil
cap_add | Capabilities to add to container | String, Array | nil
cidfile | File to store container ID | String | nil
container_name | Name for container/service | String | nil
cookbook | Cookbook to grab any templates | String | docker
cpu_shares | CPU shares for container | Fixnum | nil
detach | Detach from container when starting | TrueClass, FalseClass | nil
device | Device(s) to pass through to container | String, Array | nil
dns | DNS servers for container | String, Array | nil
dns_search | DNS search domains for container | String, Array | nil
entrypoint | Overwrite the default entrypoint set by the image | String | nil
env | Environment variables to pass to container | String, Array | nil
env_file | Read in a line delimited file of ENV variables | String | nil
expose | Expose a port from the container without publishing it to your host | Fixnum, String, Array | nil
hostname | Container hostname | String | nil
image | Image for container | String | LWRP name
init_type | Init type for container service handling | FalseClass, String | `node['docker']['container_init_type']`
init_template | Template to use for init configuration | String | nil
link | Add link to another container | String, Array | nil
label | Options to pass to underlying labeling system | String | nil
lxc_conf | Custom LXC options | String, Array | nil
memory | Set memory limit for container | Fixnum, String | nil
net | [Configure networking](http://docs.docker.io/reference/run/#network-settings) for container | String | nil
networking (*DEPRECATED*) | Configure networking for container | TrueClass, FalseClass | true
opt | Custom driver options | String, Array | nil
port | Map network port(s) to the container | Fixnum (*DEPRECATED*), String, Array | nil
privileged | Give extended privileges | TrueClass, FalseClass | nil
public_port (*DEPRECATED*) | Map host port to container | Fixnum | nil
publish_exposed_ports | Publish all exposed ports to the host interfaces | TrueClass, FalseClass | false
remove_automatically | Automatically remove the container when it exits (incompatible with detach) | TrueClass, FalseClass | false
restart | Restart policy for the container (no, on-failure, always) | String | nil
socket_template | Template to use for configuring socket (relevent for init_type systemd only) | String | nil
stdin | Attach container's stdin | TrueClass, FalseClass | nil
tty | Allocate a pseudo-tty | TrueClass, FalseClass | nil
user | User to run container | String | nil
volume | Create bind mount(s) with: [host-dir]:[container-dir]:[rw\|ro]. If "container-dir" is missing, then docker creates a new volume. | String, Array | nil
volumes_from | Mount all volumes from the given container(s) | String | nil
working_directory | Working directory inside the container | String | nil

Run a container:

```ruby
docker_container 'myImage' do
  detach true
end
```

Run a container via command:

```ruby
docker_container 'busybox' do
  command 'sleep 9999'
  detach true
end
```

Run a container from image (docker-registry for example):

```ruby
docker_container 'docker-registry' do
  image 'samalba/docker-registry'
  detach true
  hostname 'docker-registry.example.com'
  port '5000:5000'
  env 'SETTINGS_FLAVOR=local'
  volume '/mnt/docker:/docker-storage'
end
```

#### docker_container action :start

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
attach | Attach container's stdout/stderr and forward all signals to the cookbook | Cookbook to grab any templates | String | docker
init_type | Init type for container service handling | FalseClass, String | `node['docker']['container_init_type']`
init_template | Template to use for init configuration | String | nil
socket_template | Template to use for configuring socket (relevent for init_type systemd only) | String | nil
stdin | Attach container's stdin | TrueClass, FalseClass | nil

Start a stopped container:

```ruby
docker_container 'shipyard' do
  action :start
end
```

#### docker_container action :stop

These attributes are associated with this LWRP action.

Attribute | Description | Type | Default
----------|-------------|------|--------
cookbook | Cookbook to grab any templates | String | docker
init_type | Init type for container service handling | FalseClass, String | `node['docker']['container_init_type']`
init_template | Template to use for init configuration | String | nil

CHANGELOG.md
## 0.52.2

* Increase the Docker daemon timeout

## 0.51.0

* Rename helper class to avoid collisions with mainline docker cookbook

## 1.0.0 (unreleased)

* Changed default Docker port to IANA assigned 2375

Attribute deprecations:

* storage_type attribute renamed to storage_driver
* virtualization_type attribute renamed to exec_driver
* image LWRP dockerfile, image_url, and path attributes replaced with source attribute
* container LWRP Fixnum port attribute: use full String notation from Docker documentation in port attribute instead
* container LWRP public_port attribute: use port attribute instead
* container LWRP networking attribute: use net attribute instead

## 0.37.0

Please note some important changes with this release:

* The sysconfig DOCKER_OPTS improvement in [#250][] can potentially change the behavior of that environment variable as it now allows shell interpolation of any embedded variables. This should not affect most environments. If your DOCKER_OPTS does contains any expected `$`, please escape via `\$` for previous behavior or be sure it will behave as expected before upgrading.
* The daemon restart option (which is deprecated) has been defaulted to `nil` instead of `false` when `node['docker']['container_init_type']` is set to prevent issues with container restart policies. If you're dependent on the daemon option, please be sure to update your `node['docker']['restart']`  appropriately.
* This release also defaults systemd docker host to `fd://` to match upstream, enabling socket activation properly. Adjust `node['docker']['host']` if necessary.

* Bugfix: [#239][]: Upstart: install inotify tools only once (avoid CHEF-3694 warning) (thanks jperville)
* Bugfix: [#240][]: Fixed dead service containers not being restarted on docker_container :run (thanks jperville)
* Bugfix: [#244][]: Made docker_container action :remove remove the actual upstart service file (thanks jperville)
* Bugfix: [#246][]: Lengthen shell_out timeout as workaround for slow docker_container action stop (thanks jperville)
* Bugfix: [#258][]: Fix checking docker container status on debian (thanks fxposter)
* Bugfix: [#260][]: Fix accidental port changing when using systemd templates (thanks fxposter)
* Bugfix: [#266][]: Get tests working on master (thanks tduffield)
* Bugfix: [#267][]: Replace outdated testcontainerd (thanks tduffield)
* Bugfix: [#269][]: Fix tests on Travis by following Rubocop style guidelines (container LWRP) (thanks fxposter)
* Bugfix: [#280][] / [#281][]: Fix port handling when omitted in container LWRP (thanks databus23)
* Bugfix: [#284][] / [#285][]: runit finish script to stop a container (thanks xmik)
* Bugfix: [#288][]: Fix docker.socket unit for RHEL7 (thanks databus23)
* Bugfix: [#292][]: readme formatting fix (thanks wormzer)
* Improvement: [#208][]: Add CentOS/RHEL 7 support (thanks dermusikman and intoximeters)
* Improvement: [#232][]: Added support for insecure-registry docker daemon option (thanks jperville)
* Improvement: [#233][] / [#234][]: Added support for registry-mirror docker daemon option (thanks jperville and tarnfeld)
* Improvement: [#237][]: Deprecate the restart daemon option (thanks jperville)
* Improvement: [#238][]: Added docker_container restart attribute (thanks jperville)
* Improvement: [#242][]: Added docker_container action :create (thanks jperville)
* Improvement: [#245][]: Add a Gitter chat badge to README.md (thanks tduffield)
* Improvement: [#250][]: Use double-quotes for DOCKER_OPTS (thanks rchekaluk)
* Improvement: [#259][]: Use registry on image inspection (thanks fxposter)
* Improvement: [#263][]: Add additional_host attribute to container resource (thanks fxposter)
* Improvement: [#264][] / [#265][]: Access keyserver.ubuntu.com on port 80 (thanks sauraus)
* Improvement: [#268][]: Updated the /etc/init/docker.conf template (thanks jperville)
* Improvement: [#276][]: Added support for docker options device and cap-add (thanks hvolkmer)
* Improvement: [#279][]: Allow docker_container memory to have String value (eg. memory='1G') (thanks jperville)
* Improvement: [#287][]: redhat 7 does not need the epel repository (thanks databus23)
* Improvement: [#289][]: Update systemd service/socket files (from upstream) (thanks databus23)
* Improvement: [#296][]: Default systemd to fd:// as well as use upstream MountFlags=slave and LimitCORE=infinity
* Improvement: [#297][]: Update docker daemon SysV init scripts with upstream improvements
* Improvement: [#298][]: Further deprecate daemon restart flag by default, which interferes with restart policies

## 0.36.0
* Bugfix: [#181][]: Fixed remove_link action (thanks jperville).
* Bugfix: [#185][]: Fix for non idempotent run action on docker_container (thanks bplunkert).
* Bugfix: [#188][]: Applied temporary workaround to address the libcgmanager error to users running LXC on Ubuntu 14.04.
* Bugfix: [#196][]: Address Helpers module naming conflict (thanks sethrosenblum).
* Bugfix: [#200][]: Fix how service actions are handled by docker_container resource (thanks brianhartsock).
* Bugfix: [#202][]: Correctly check for the kernel.release version on Debian (thanks Tritlo, paweloczadly).
* Bugfix: [#203][]: Fix pull notifications for tagged images (thanks hobofan).
* Bugfix: [#205][]: Fix current_resource.name assignments in docker_container provider (thanks jperville).
* Bugfix: [#206][]: Fixes to container name detection (thanks jperville).
* Enhancement: [#217][]: Explicitly set key and keyserver for docker apt repository (thanks sethrosenblum).
* Improvement: Pull in init script changes from upstream for sysv and systemd.
* Bugfix: [#219][]: Explicitly set Upstart provider for Ubuntu 14.04 and 14.10 (thanks methodx).
* Improvement: [#220][]: Create graph directory if it is specified (thanks jontg).
* Bugfix: [#224][]: Fix runit container template to properly use exec (thanks waisbrot).
* Bugfix: Appropriately check for LXC when using the binary recipe on Fedora.

attributes/default.rb
include_attribute 'golang'

# Installation/System attributes

default['docker']['arch'] =
  case node['kernel']['machine']
  when 'x86_64' then 'x86_64'
  # If Docker ever supports 32-bit or other architectures
  # when %r{i[3-6]86} then 'i386'
  else 'x86_64'
  end
default['docker']['group_members'] = []
default['docker']['init_type'] = value_for_platform(
  %w(amazon debian oracle) => {
    'default' => 'sysv'
  },
  %w(redhat centos) => {
    %w(6.0 6.1 6.2 6.3 6.4 6.5 6.6) => 'sysv',
    'default' => 'systemd'
  },
  %w(fedora) => {
    'default' => 'systemd'
  },
  %w(ubuntu) => {
    'default' => 'upstart'
  },
  'default' => 'upstart'
)
default['docker']['install_type'] = value_for_platform(
  %w(centos debian fedora mac_os_x redhat ubuntu amazon) => {
    'default' => 'package'
  },
  'default' => 'binary'
)
default['docker']['install_dir'] =
  case node['docker']['install_type']
  when 'binary' then '/usr/local/bin'
  when 'source' then node['go']['gobin']
  else '/usr/bin'
  end
default['docker']['ipv4_forward'] = true
default['docker']['ipv6_forward'] = true
default['docker']['logfile'] = nil
default['docker']['version'] = nil

# Actions: :warn, :fatal
default['docker']['alert_on_error_action'] = :fatal

## Binary installation attributes

default['docker']['binary']['dependency_packages'] = value_for_platform_family(
  'debian' => %w(procps xz-utils),
  'rhel' => %w(procps xz),
  'default' => %w()
)
default['docker']['binary']['version'] = node['docker']['version'] || 'latest'
default['docker']['binary']['checksum'] =
case node['kernel']['name']
when 'Darwin'
  case node['docker']['binary']['version']
  when '0.10.0' then '416835b2e83e520c3c413b4b4e4ae34bca20704f085b435f4c200010dd1ac3b7'
  when '0.11.0' then '9db839b56a8656cfcef1f6543e9f75b01a774fdd6a50457da20d8183d6b415fa'
  when '0.11.1' then '386ffa26e52856107efb0b3075625d5b2331fa5acc8965fef87c1ab7d900c4e9'
  when '0.12.0' then 'a38dccb7f544fad4ef2f95243bef7e2c9afbd76de0e4547b61b27698bf9065f3'
  when '1.0.0' then '67c3c9f285584533ac365a56515f606fc91d4dcd0bfa69c2f159eeb5e37ea3b8'
  when '1.0.1' then 'b662e7718f0a8e23d2e819470a368f257e2bc46f76417712360de7def775e9d4'
  end
when 'Linux'
  case node['docker']['binary']['version']
  when '0.10.0' then 'ce1f5bc88a99f8b2331614ede7199f872bd20e4ac1806de7332cbac8e441d1a0'
  when '0.11.0' then 'f80ba82acc0a6255960d3ff6fe145a8fdd0c07f136543fcd4676bb304daaf598'
  when '0.11.1' then 'ed2f2437fd6b9af69484db152d65c0b025aa55aae6e0991de92d9efa2511a7a3'
  when '0.12.0' then '0f611f7031642a60716e132a6c39ec52479e927dfbda550973e1574640135313'
  when '1.0.0' then '55cf74ea4c65fe36e9b47ca112218459cc905ede687ebfde21b2ba91c707db94'
  when '1.0.1' then '1d9aea20ec8e640ec9feb6757819ce01ca4d007f208979e3156ed687b809a75b'
  end
end
default['docker']['binary']['url'] = "http://get.docker.io/builds/#{node['kernel']['name']}/#{node['docker']['arch']}/docker-#{node['docker']['binary']['version']}"

## Package installation attributes

default['docker']['package']['action'] = 'install'
default['docker']['package']['distribution'] = value_for_platform(
  'ubuntu' => {
    '14.04' => 'ubuntu-trusty'
  },
  'default' => 'docker'
)
default['docker']['package']['name'] = value_for_platform(
  'amazon' => {
    'default' => 'docker'
  },
  %w(centos redhat) => {
    %w(6.0 6.1 6.2 6.3 6.4 6.5 6.6) => 'docker-io',
    'default' => 'docker'
  },
  'fedora' => {
    'default' => 'docker-io'
  },
  'debian' => {
    'default' => 'lxc-docker'
  },
  'mac_os_x' => {
    'default' => 'homebrew/binary/docker'
  },
  'ubuntu' => {
    %w(12.04 12.10 13.04 13.10 14.04) => 'docker-engine',
    'default' => 'docker.io'
  },
  'default' => nil
)
default['docker']['package']['repo_url'] = value_for_platform(
  'debian' => {
    'default' => 'https://get.docker.io/ubuntu'
  },
  'ubuntu' => {
    %w(12.04 12.10 13.04 13.10 14.04) => 'https://apt.dockerproject.org/repo',
    'default' => nil
  },
  'default' => nil
)
default['docker']['package']['repo_keyserver'] = 'hkp://pgp.mit.edu:80'
# Found at https://get.docker.io/ubuntu/
default['docker']['package']['repo_key'] = '58118E89F3A912897C070ADBF76221572C52609D'

## Source installation attributes

default['docker']['source']['ref'] = 'master'
default['docker']['source']['url'] = 'https://github.com/dotcloud/docker.git'

# Docker Daemon attributes

default['docker']['api_enable_cors'] = nil

# DEPRECATED: will be removed in chef-docker 1.0
default['docker']['bind_socket'] = nil
# DEPRECATED: will be removed in chef-docker 1.0
default['docker']['bind_uri'] = nil

default['docker']['bip'] = nil
default['docker']['bridge'] = nil
default['docker']['debug'] = nil
default['docker']['dns'] = nil
default['docker']['dns_search'] = nil
default['docker']['exec_driver'] = nil

# DEPRECATED: will be removed in chef-docker 1.0
default['docker']['virtualization_type'] = node['docker']['exec_driver']

default['docker']['graph'] = nil
default['docker']['group'] = node['docker']['group_members'].empty? ? nil : 'docker'

# DEPRECATED: Support for bind_socket/bind_uri
default['docker']['host'] =
  if node['docker']['bind_socket'] || node['docker']['bind_uri']
    Array(node['docker']['bind_socket']) + Array(node['docker']['bind_uri'])
  elsif node['docker']['init_type'] == 'systemd'
    'fd://'
  else
    'unix:///var/run/docker.sock'
  end
default['docker']['http_proxy'] = nil
default['docker']['icc'] = nil
default['docker']['insecure-registry'] = nil
default['docker']['ip'] = nil
default['docker']['iptables'] = nil
default['docker']['mtu'] = nil
default['docker']['no_proxy'] = nil
default['docker']['options'] = nil
default['docker']['pidfile'] = nil
default['docker']['ramdisk'] = false
default['docker']['registry-mirror'] = nil
default['docker']['selinux_enabled'] = nil
default['docker']['storage_driver'] = nil
default['docker']['storage_opt'] = nil

# DEPRECATED: will be removed in chef-docker 1.0
default['docker']['storage_type'] = node['docker']['storage_driver']

default['docker']['tls'] = nil
default['docker']['tlscacert'] = nil
default['docker']['tlscert'] = nil
default['docker']['tlskey'] = nil
default['docker']['tlsverify'] = nil
default['docker']['tmpdir'] = nil

# LWRP attributes

default['docker']['docker_daemon_timeout'] = 20

## docker_container attributes

default['docker']['container_cmd_timeout'] = 60
default['docker']['container_init_type'] = node['docker']['init_type']

## docker_image attributes

default['docker']['image_cmd_timeout'] = 300

## docker_registry attributes

default['docker']['registry_cmd_timeout'] = 60

# Other attributes

# DEPRECATED: will be removed in chef-docker 1.0
default['docker']['restart'] = nil
